crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -643,39 +649,59 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
         let mut plain_accounts_cursor = self.tx.cursor_write::<tables::PlainAccountState>()?;
         let mut plain_storage_cursor = self.tx.cursor_dup_write::<tables::PlainStorageState>()?;
 
-        let mut reverts: RevertsInit = HashMap::new();
-
-        // add account changeset changes
+        // Iterate over the account changesets in reverse order.
         for (block_number, account_before) in account_changeset.into_iter().rev() {
+            // Destructure the account changeset to get the old info and address.
             let AccountBeforeTx { info: old_info, address } = account_before;
-            match state.entry(address) {
-                hash_map::Entry::Vacant(entry) => {
-                    let new_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
-                    entry.insert((old_info, new_info, HashMap::new()));
-                }
-                hash_map::Entry::Occupied(mut entry) => {
-                    // overwrite old account state.
-                    entry.get_mut().0 = old_info;
+
+            // If the address is not already in the bundle builder's state, add it.
+            if !bundle_builder.get_states().contains(&address) {
+                // Seek the address in the plain account state.
+                if let Some(new_info) = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1) {
+                    // Add the new info to the bundle builder.
+                    bundle_builder =
+                        bundle_builder.state_present_account_info(address, new_info.into());
                 }
             }
-            // insert old info into reverts.
-            reverts.entry(block_number).or_default().entry(address).or_default().0 = Some(old_info);
+
+            // Add the old account info to the bundle builder if it exists.
+            if let Some(old_info) = old_info {
+                bundle_builder =
+                    bundle_builder.state_original_account_info(address, old_info.into());
+            }
```

Comment:
review from https://github.com/paradigmxyz/reth/pull/8264/files#r1646535481 @rkrasiuk 

> previous behavior: if the entry is occuppied, overwrite the original account state even if old_info is None
current behavior: only overwrite the original account state if old_info is Some. 
if old_info is None, original info must still be overwritten

the methods `BundleBuilder::state_original_account_info` and `BundleBuilder::state_present_account_info` don't take an `Option<AccountInfo` as param. so I'm wondering:

- if entry is occupied but the original info is `None`, should we pass the `new_info` to `BundleBuilder::state_original_account_info`?
- there is another method `BundleBuilder::set_revert_account_info`, that takes `Option<Option<AccountInfo>>`, is that the one we should be calling here?

Reply:
@DaniPopes, you're probably familiar with the revm api and can answer this too?

Reply:
I'm not familiar with state/transitions/bundles APIs